rfd/0100-proxy-ssh-grpc.md

```diff
@@ -0,0 +1,197 @@
+message which declares the target server that the connection is for. If the
+target exists and session control allows, the server will establish the
+connection and respond with a `ConnectionDetails` message.
+
+After the initial ceremony, either side of the connection may transmit `Frame`s
+until the connection terminates. Each `Frame` includes the raw payload and
+the `Protocol` which generated the payload to allow for multiple protocols to
+leverage the same stream as a transport.
+
+The most pressing need for multiplexed protocols is to ensure that `tsh ssh` functions
+properly in proxy recording mode. Since the Proxy creates an SSH connection to the Node
+on behalf of the user in proxy recording mode the user *must* forward their agent to
+facilitate the connection. Currently when `tsh` determines the Proxy is performing the
+session recording it will forward the user's agent over a SSH channel. The Proxy then
+communicates SSH Agent protocol over that channel to sign requests. `tsh` utilizes
+`agent.ForwardToAgent` and `agent.RequestAgentForwarding` from `x/crypto/ssh/agent` to
+set up the channel and serve the agent over the channel to the Proxy.
+
+To achieve the same functionality using the gRPC stream proposed above, the SSH Agent protocol can
+be multiplexed over the stream in addition to the SSH protocol. When `tsh` determines proxy
+recording is in effect it can leverage `agent.ServeAgent` directly, passing in an `io.ReadWriter`
+which sends and receives `Frame`s with `Protocol_SSH_AGENT`populated in the protocol field when
+it is written to and read from. The server side can communicate with the local agent by using
+`agent.NewClient` on a similar `io.ReadWriter`.
+
+The end result is both SSH and SSH Agent protocol being transported across the same stream
+to enable both the SSH connection to the target Node and allowing the Proxy to communicate
+with the user's local SSH agent in a similar manner to way it works to date.
+
+## Performance
+
+Below are two traces captured with both Proxy transport mechanisms that illustrate the latency
+reduction.
+
+#### SSH
+![SSH Transport](assets/0100-ssh-transport.png)
+#### gRPC
+![gRPC Transport](assets/0100-grpc-transport.png)
+
+
+The existing SSH transport took 6.73s to execute `tsh user@foo uptime`, while the same
+command via the gRPC transport took 5.36s resulting in a ~20% reduction in latency. 
+
+## Security
+
+The gRPC server will require mTLS for authentication and perform the same RBAC
+and session control checks as the current SSH server does. Agent forwarding will
+occur as it does today with the exception that the SSH Agent Protocol will use a
+gRPC stream instead of an SSH channel for transport.
+
+## UX
+
+The behavior of `tsh ssh` should remain the same regardless of the configured
+session recording mode. The time it takes to establish a session may be noticeably
+faster depending on proximity of the client and the Proxy.
```

Comment:
Do you plan to add a fallback to the old double handshake method if the ALPN connection fails? Technically we don't support newer clients with older servers, but in practice it usually works.

Reply:
+1 for this, we can't expect users to micromanage their `tsh` version, IMO.